agent/grpc/resolver/resolver.go

```diff
@@ -0,0 +1,287 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/consul/agent/router"
+	"google.golang.org/grpc/resolver"
+)
+
+var registerLock sync.Mutex
+
+// RegisterWithGRPC registers the ServerResolverBuilder as a grpc/resolver.
+// This function exists to synchronize registrations with a lock.
+// grpc/resolver.Register expects all registration to happen at init and does
+// not allow for concurrent registration. This function exists to support
+// parallel testing.
+func RegisterWithGRPC(b *ServerResolverBuilder) {
+	registerLock.Lock()
+	defer registerLock.Unlock()
+	resolver.Register(b)
+}
+
+// Nodes provides a count of the number of nodes in the cluster. It is very
+// likely implemented by serf to return the number of LAN members.
+type Nodes interface {
+	NumNodes() int
+}
+
+// ServerResolverBuilder tracks the current server list and keeps any
+// ServerResolvers updated when changes occur.
+type ServerResolverBuilder struct {
+	// datacenter of the local agent.
+	datacenter string
+	// scheme used to query the server. Defaults to consul. Used to support
+	// parallel testing because gRPC registers resolvers globally.
+	scheme string
+	// servers is an index of Servers by Server.ID
+	// TODO: this assumes that server.ID is unique across all datacenter. Is that
+	// a safe assumption?
+	servers map[string]*metadata.Server
+	// resolvers is an index of connections to the serverResolver which manages
+	// addresses of servers for that connection.
+	resolvers map[resolver.ClientConn]*serverResolver
+	// nodes provides the number of nodes in the cluster.
+	nodes Nodes
+	// lock for servers and resolvers.
+	lock sync.RWMutex
+}
+
+var _ resolver.Builder = (*ServerResolverBuilder)(nil)
+
+type Config struct {
+	// Datacenter of the local agent.
+	Datacenter string
+	// Scheme used to connect to the server. Defaults to consul.
+	Scheme string
+}
+
+func NewServerResolverBuilder(cfg Config, nodes Nodes) *ServerResolverBuilder {
+	if cfg.Scheme == "" {
+		cfg.Scheme = "consul"
+	}
+	return &ServerResolverBuilder{
+		scheme:     cfg.Scheme,
+		datacenter: cfg.Datacenter,
+		nodes:      nodes,
+		servers:    make(map[string]*metadata.Server),
+		resolvers:  make(map[resolver.ClientConn]*serverResolver),
+	}
+}
+
+// Run periodically reshuffles the order of server addresses  within the
+// resolvers to ensure the load is balanced across servers.
+//
+// TODO: this looks very similar to agent/router.Manager.Start, which is the
+// only other caller of ComputeRebalanceTimer. Are the values passed to these
+// two functions different enough that we need separate goroutines to rebalance?
+// or could we have a single thing handle the timers, and call both rebalance
+// functions?
```

Comment:
Calling out this TODO as well. I haven't traced all the code to understanding if the values passed to `ComputeRebalanceTimer` are different or not.

Looking at it again, this seems to balance based on the values for the local DC, but the resolvers could be for all DCs. Do we really need this dependency on `router.ComputeRebalanceTimer` or could we use our own values and the shared stuff from `lib` ?

Reply:
I created a patch to demonstrate this: https://github.com/hashicorp/consul/commit/be0ce16557e3e50f8a88b84a189c81da2fe7d310

I believe this is more correct because the rebalance happens based on the values for a specific DC. Currently the timer value is being calculated for all servers.

Reply:
Your `Rebalancer` patch looks like it would work and is probably a little cleaner although it requires some plumbing changes (which you did). I'm not sure saving one timer goroutine is enough of an optimization on it's own but I think the overall structure of the code is a little cleaner that way.

> I believe this is more correct because the rebalance happens based on the values for a specific DC. 

I'm finding it a bit hard to understand the difference from your patch!

The rationale in this PR for why we need to filter the number of servers to only ones in the current (or target) DC is that client agents only ever talk to their local servers, even if they are sending a request to a remote DC, that is sent to their local server and proxied from there. So client load (which is what we are trying to balance here) should all be being worked out based on the local DC only.

That said, I can't see how in your patch the behaviour is actually different, assuming that this line:

```go
func (m *Manager) refreshServerRebalanceTimer() time.Duration {
	l := m.getServerList()
```

Is only returning local DC servers. If that line is returning _all_ the servers then I'd say this is actually a bug that's probably existed for a long time - the intent there is to spread/balance the local DC clients over the local DC servers without causing too many reconnections in any given period, if it's using the number of servers from all DCs then it's not working as expected (but probably not that far out most of the time).

Or am I missunderstanding?

Reply:
Yeah I checked it out and each Manager is only the set of servers in one DC so `m.getServerList()` is only the local DCs above.

So I think that's equivalent. The reason we needed to explicitly filter by DC in this PR was that there is only one gRPC client per process which means on a server which might be making upstream gRPC calls to multiple sets of servers in other DCs this same Router object has to maintain all of those sets. But it should continue to use the same rebalancing logic for each one - based only on the number of servers in that DC and the number of WAN clients trying to talk to it.

Client agents should never connect to more than one set of servers so it doesn't matter there, and in practice WAN pools are never going to be big enough that the cluster scaling maths makes any differece (you'd need thousands of server nodes across hundreds of DCs before that were true) so it's probably not a huge deal in practice.

But we really do want to ensure that client -> server balancing in the local DC has equivalent balancing logic to now.

Reply:
Ah right! I was thinking that the problem was around the WAN case, because all resolvers get rebalanced at the same time, based on counts for LAN nodes. However that is not changed in be0ce16557e3e50f8a88b84a189c81da2fe7d310 because all the managers call the same function. If the managers were to pass the DC name, and `Rebalance()` only did so for resolvers in that DC, then the behaviour would be the same as we have today for connections.

I think in that patch clients are being rebalanced more often than we expect because the manager for each DC is also calling rebalance.

I was not worried about the extra goroutine and timer. I think the advantage to that patch is less coupling between components, and less code doing roughly the same thing.

I'll fix the patch to pass the DC name, so that we only rebalance the resolvers for the correct DC, and apply it to this PR.